feat(metrics): Add opencost_build_info metric

Pokom opened the pull request:
## What does this PR change?
* Adds a `opencost_build_info` metric with a `version` label

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
* Limited impact, will add one more metric

## Does this PR address any GitHub or Zendesk issues?
* Closes #1576 

## How was this PR tested?
* Locally

## Does this PR require changes to documentation?
* N/A

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?


I'm looking for some input from other maintainers on:
- How to best get the version information(Git SHA, ldflags set a build time etc..)
- Is this the right place into the codebase to set the build information?

Pokom committed:
feat(metrics): Add opencost_build_info metric

Signed-off-by: pokom <mark.poko@grafana.com>

Pokom committed:
feat(build): Add version and revision as labels

This adds a module `pkg/build` that is responsible for setting
information about the build and providing a method `GetBuildInfo` that
will return an object with associated build information.

The way that it works is at build time, we need to provide a set of
[extldflags](https://pkg.go.dev/cmd/link) which will set the variables defined
in `pkg/build/build.go`. There is an associated Makefile that will
populate the following variables:
- IMAGE_TAG => Version
- GIT_REVISION => Revision
- GIT_BRANCH => Branch
- GIT_LAST_COMMIT_DATE => BuildDate

These are then exported via the `opencost_build_info` metric as labels.

Signed-off-by: pokom <mark.poko@grafana.com>

Pokom committed:
Merge branch 'develop' into feat/build-info-metric

AjayTripathy commented:
https://github.com/opencost/opencost/blob/v1.99/pkg/env/costmodelenv.go#L170 Just a note that the env variable of published version seems relevant!

Pokom committed:
Merge branch 'develop' into feat/build-info-metric

Pokom committed:
feat(metrics): Migrate to using existing version.go

Delete the `build.go` file which was redundent to
`pkg/version/version.go`.

Update Makefile to eliminate redundent information.

Signed-off-by: pokom <mark.poko@grafana.com>

Pokom commented:
Thanks for the nudge @AjayTripathy! After looking through, I think the two version variables *should* be the same, but I am hesitant to rely upon one that can be overridden by users for the `opencost_build_info` metric. I feel it would be pretty easy to set that environment variable at runtime(IE, helm/jsonnet/manually) and then forget about it. 

Out of curiosity, what is the use case for being able to override `APP_VERSION`?

Pokom commented:
@AjayTripathy I think this PR is ready for review at this point. It should be good enough for a first pass to start exporting the version which will be enough for me to setup rolling out `opencost` in an automated fashion. Let me know if there's anything you'd like to see tested prior to a review.

AjayTripathy commented:
SGTM will take a look @Pokom ! Just curious how are you planning to automate the rollouts?

AjayTripathy committed:
Merge branch 'develop' into feat/build-info-metric

AjayTripathy commented on pkg/metrics/telemetry.go:
```diff
@@ -42,7 +51,7 @@ func InitKubecostTelemetry(config *MetricsConfig) {
 		}, []string{"handler", "method", "code"})
 
-		prometheus.MustRegister(requestsCount, responseTime, responseSize)
+		prometheus.MustRegister(requestsCount, responseTime, responseSize, buildInfo)
```
Did you mean to do this on the requestsCounts metric? I don't think that's necessary.

Pokom replied:
@AjayTripathy I didn't think this impacted the other metrics. I was piggy backing off of the existing `prometheus.MustRegister` calls . If needed, I could move out buildInfo to it's own separate `prometheus.MustRegister` call

AjayTripathy replied:
Oh NVM I think it's fine.

AjayTripathy commented:
@Pokom quick question? I think the metric got accidentaly added as a label on another request?

Pokom commented:
> @AjayTripathy Just curious how are you planning to automate the rollouts?

We have internal tooling setup to periodically check if there's a new container image in the registry by querying prometheus for the specific `*_build_info` metric and then does some GitOps magic under the hood to stage a rollout. 

>  I think the metric got accidentaly added as a label on another request?

Provided my response inline. I can double check the exported metrics, but I thought that `prometheus.MustRegister` accepted a variable amount of collectors and registered them independently. I don't think it will impact the other metrics.
